Register PrimTorch refs as decompositions.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #76835

For the most part, PrimTorch refs have the same signature as their
ATen equivalents.  I modify most PrimTorch refs to register themselves
as decompositions, using the prim name they wrap to find the aten name
(except for a few cases where the prim/aten names mismatch).  There are
some exclusions, falling into one of two categories:

- The torch equivalent was already implemented as a CompositeImplicitAutograd
  decomposition in C++

- The ref doesn't support enough features (e.g., the real deal has more
  kwargs / overloads than are currently implemented)

PrimTorch refs are written as a single function that supports all
overloads, and this style is convenient for cases where we have a bundle
of overloads for what morally is a single overload with a Union type
on an argument (which we ought to have supported in
native_functions.yaml but blah); to support registering a single decomp
for all the overloads, we modify register_decomposition to register
to ALL overloads if you pass it an overload packet.  This is technically
BC breaking but no tests started failing because of it.

There's also two other major technical pieces to this PR:

* Wrapped number tensors get converted back into numbers when they go through `__torch_dispatch__`, fixing #74400 
* torch.ops now accept integers/floats anywhere Tensors were accepted, and they turn into wrapped scalar numbers
* PrimTorch refs now handle CPU scalar + CUDA correctly, fixes #76871

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Register PrimTorch refs as decompositions.

For the most part, PrimTorch refs have the same signature as their
ATen equivalents.  I modify most PrimTorch refs to register themselves
as decompositions, using the prim name they wrap to find the aten name
(except for a few cases where the prim/aten names mismatch).  There are
some exclusions, falling into one of two categories:

- The torch equivalent was already implemented as a CompositeImplicitAutograd
  decomposition in C++

- The ref doesn't support enough features (e.g., the real deal has more
  kwargs / overloads than are currently implemented)

PrimTorch refs are written as a single function that supports all
overloads, and this style is convenient for cases where we have a bundle
of overloads for what morally is a single overload with a Union type
on an argument (which we ought to have supported in
native_functions.yaml but blah); to support registering a single decomp
for all the overloads, we modify register_decomposition to register
to ALL overloads if you pass it an overload packet.  This is technically
BC breaking but no tests started failing because of it.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Register PrimTorch refs as decompositions."


For the most part, PrimTorch refs have the same signature as their
ATen equivalents.  I modify most PrimTorch refs to register themselves
as decompositions, using the prim name they wrap to find the aten name
(except for a few cases where the prim/aten names mismatch).  There are
some exclusions, falling into one of two categories:

- The torch equivalent was already implemented as a CompositeImplicitAutograd
  decomposition in C++

- The ref doesn't support enough features (e.g., the real deal has more
  kwargs / overloads than are currently implemented)

PrimTorch refs are written as a single function that supports all
overloads, and this style is convenient for cases where we have a bundle
of overloads for what morally is a single overload with a Union type
on an argument (which we ought to have supported in
native_functions.yaml but blah); to support registering a single decomp
for all the overloads, we modify register_decomposition to register
to ALL overloads if you pass it an overload packet.  This is technically
BC breaking but no tests started failing because of it.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Register PrimTorch refs as decompositions."


For the most part, PrimTorch refs have the same signature as their
ATen equivalents.  I modify most PrimTorch refs to register themselves
as decompositions, using the prim name they wrap to find the aten name
(except for a few cases where the prim/aten names mismatch).  There are
some exclusions, falling into one of two categories:

- The torch equivalent was already implemented as a CompositeImplicitAutograd
  decomposition in C++

- The ref doesn't support enough features (e.g., the real deal has more
  kwargs / overloads than are currently implemented)

PrimTorch refs are written as a single function that supports all
overloads, and this style is convenient for cases where we have a bundle
of overloads for what morally is a single overload with a Union type
on an argument (which we ought to have supported in
native_functions.yaml but blah); to support registering a single decomp
for all the overloads, we modify register_decomposition to register
to ALL overloads if you pass it an overload packet.  This is technically
BC breaking but no tests started failing because of it.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Register PrimTorch refs as decompositions."


For the most part, PrimTorch refs have the same signature as their
ATen equivalents.  I modify most PrimTorch refs to register themselves
as decompositions, using the prim name they wrap to find the aten name
(except for a few cases where the prim/aten names mismatch).  There are
some exclusions, falling into one of two categories:

- The torch equivalent was already implemented as a CompositeImplicitAutograd
  decomposition in C++

- The ref doesn't support enough features (e.g., the real deal has more
  kwargs / overloads than are currently implemented)

PrimTorch refs are written as a single function that supports all
overloads, and this style is convenient for cases where we have a bundle
of overloads for what morally is a single overload with a Union type
on an argument (which we ought to have supported in
native_functions.yaml but blah); to support registering a single decomp
for all the overloads, we modify register_decomposition to register
to ALL overloads if you pass it an overload packet.  This is technically
BC breaking but no tests started failing because of it.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

Chillee commented on torch/_refs/__init__.py:
```diff
@@ -534,7 +543,8 @@ def _ref(a: Tensor, *, out: Optional[Tensor] = None) -> Tensor:
 
 erfinv = _make_elementwise_unary_reference(
-    prims.erf_inv, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.INT_TO_FLOAT
+    prims.erf_inv, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.INT_TO_FLOAT,
```
Is there nowhere that we're mapping from references to torch ops? That seems like the better place to automatically map between ATen and reference ops.

ezyang replied:
The mapping is implicit in the `erfinv = ` assignment. I suppose I could do some `locals()` tricks but this seemed... less invasive.

ngimel replied:
I'm taking a note to fix this name mismatch, there's no reason to have it.

Chillee commented on torch/_decomp/__init__.py:
```diff
@@ -34,13 +34,17 @@ def decomposition_decorator(f):
             else:
-                op_overload = aten_op
-            registry[op_overload] = f
+                assert isinstance(aten_op, torch._ops.OpOverloadPacket)
```
This changes the behavior of existing decomposition registrations right? We now need to add `.default` to all of them?

ezyang replied:
Yes, unless you wanted all the overloads to be hit by your decomp.

Chillee replied:
Are you going to do that in this PR?

ezyang replied:
I can, although no tests started failing, so it seems like it was OK to hit all overloads XD

Chillee replied:
I suspect it's because not many operators fulfill both 1. have multiple non-CompositeImplicit operators, 2. have different semantics for each of them lol.

Chillee reviewed (approved):
LGTM! Some minor comments/confusions.

ezyang committed:
Update on "Register PrimTorch refs as decompositions."


For the most part, PrimTorch refs have the same signature as their
ATen equivalents.  I modify most PrimTorch refs to register themselves
as decompositions, using the prim name they wrap to find the aten name
(except for a few cases where the prim/aten names mismatch).  There are
some exclusions, falling into one of two categories:

- The torch equivalent was already implemented as a CompositeImplicitAutograd
  decomposition in C++

- The ref doesn't support enough features (e.g., the real deal has more
  kwargs / overloads than are currently implemented)

PrimTorch refs are written as a single function that supports all
overloads, and this style is convenient for cases where we have a bundle
of overloads for what morally is a single overload with a Union type
on an argument (which we ought to have supported in
native_functions.yaml but blah); to support registering a single decomp
for all the overloads, we modify register_decomposition to register
to ALL overloads if you pass it an overload packet.  This is technically
BC breaking but no tests started failing because of it.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Register PrimTorch refs as decompositions."


For the most part, PrimTorch refs have the same signature as their
ATen equivalents.  I modify most PrimTorch refs to register themselves
as decompositions, using the prim name they wrap to find the aten name
(except for a few cases where the prim/aten names mismatch).  There are
some exclusions, falling into one of two categories:

- The torch equivalent was already implemented as a CompositeImplicitAutograd
  decomposition in C++

- The ref doesn't support enough features (e.g., the real deal has more
  kwargs / overloads than are currently implemented)

PrimTorch refs are written as a single function that supports all
overloads, and this style is convenient for cases where we have a bundle
of overloads for what morally is a single overload with a Union type
on an argument (which we ought to have supported in
native_functions.yaml but blah); to support registering a single decomp
for all the overloads, we modify register_decomposition to register
to ALL overloads if you pass it an overload packet.  This is technically
BC breaking but no tests started failing because of it.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Register PrimTorch refs as decompositions."


For the most part, PrimTorch refs have the same signature as their
ATen equivalents.  I modify most PrimTorch refs to register themselves
as decompositions, using the prim name they wrap to find the aten name
(except for a few cases where the prim/aten names mismatch).  There are
some exclusions, falling into one of two categories:

- The torch equivalent was already implemented as a CompositeImplicitAutograd
  decomposition in C++

- The ref doesn't support enough features (e.g., the real deal has more
  kwargs / overloads than are currently implemented)

PrimTorch refs are written as a single function that supports all
overloads, and this style is convenient for cases where we have a bundle
of overloads for what morally is a single overload with a Union type
on an argument (which we ought to have supported in
native_functions.yaml but blah); to support registering a single decomp
for all the overloads, we modify register_decomposition to register
to ALL overloads if you pass it an overload packet.  This is technically
BC breaking but no tests started failing because of it.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
@Chillee @jansel  one thing that I'm a little concerned about is that turning this on by default puts pressure on users of decompositions to say whether or not they want the primtorch decomps or not. As of this PR, you're just always going to get the primtorch decomp. An unprincipled addition we could make is a boolean flag to let AOTAutograd users decide if they want them or not. 

There's also a short term problem, which is that if you AOTAutograd trace through the primtorch decomps, they won't actually show up in the graph as prims, they'll show up as the torch operators that the primtorch things are implemented with. I know how to fix this bug I need to actually register the prims. Not entirely sure about ordering.

ezyang committed:
Update on "Register PrimTorch refs as decompositions."


For the most part, PrimTorch refs have the same signature as their
ATen equivalents.  I modify most PrimTorch refs to register themselves
as decompositions, using the prim name they wrap to find the aten name
(except for a few cases where the prim/aten names mismatch).  There are
some exclusions, falling into one of two categories:

- The torch equivalent was already implemented as a CompositeImplicitAutograd
  decomposition in C++

- The ref doesn't support enough features (e.g., the real deal has more
  kwargs / overloads than are currently implemented)

PrimTorch refs are written as a single function that supports all
overloads, and this style is convenient for cases where we have a bundle
of overloads for what morally is a single overload with a Union type
on an argument (which we ought to have supported in
native_functions.yaml but blah); to support registering a single decomp
for all the overloads, we modify register_decomposition to register
to ALL overloads if you pass it an overload packet.  This is technically
BC breaking but no tests started failing because of it.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
@albanD for the wrapped Python number changes

Chillee commented:
> As of this PR, you're just always going to get the primtorch decomp.

This seems fine? None of the current users of decompositions indiscriminately use decomps (maybe torch-mlir?). There's plenty of decomps in there that could be problematic if you can't fuse them (like say, `diagonal_backward`).

ngimel commented on torch/_refs/__init__.py:
```diff
@@ -637,6 +651,12 @@ def _ref(
 
+        # Special case CPU scalar tensors to be eligible for device transfer
+        if isinstance(a, TensorLike) and a.device.type == 'cpu' and len(a.shape) == 0:
+            a = a.item()
```
lol we are having thoughtful discussion about where and if to wrap scalar args, and here turns out we need to unwrap them.

ezyang replied:
Unfortunately it doesn't work lol

```
  File "/private/home/ezyang/pytorch-tmp/torch/_refs/__init__.py", line 663, in _ref
    result = prim(a, b)
  File "/private/home/ezyang/pytorch-tmp/torch/_prims/__init__.py", line 174, in _prim
    return impl_aten(*args, **kwargs)
TypeError: nextafter(): argument 'input' (position 1) must be Tensor, not float
```

ezyang committed:
Update on "Register PrimTorch refs as decompositions."


For the most part, PrimTorch refs have the same signature as their
ATen equivalents.  I modify most PrimTorch refs to register themselves
as decompositions, using the prim name they wrap to find the aten name
(except for a few cases where the prim/aten names mismatch).  There are
some exclusions, falling into one of two categories:

- The torch equivalent was already implemented as a CompositeImplicitAutograd
  decomposition in C++

- The ref doesn't support enough features (e.g., the real deal has more
  kwargs / overloads than are currently implemented)

PrimTorch refs are written as a single function that supports all
overloads, and this style is convenient for cases where we have a bundle
of overloads for what morally is a single overload with a Union type
on an argument (which we ought to have supported in
native_functions.yaml but blah); to support registering a single decomp
for all the overloads, we modify register_decomposition to register
to ALL overloads if you pass it an overload packet.  This is technically
BC breaking but no tests started failing because of it.

There's also two other major technical pieces to this PR:

* Wrapped number tensors get converted back into numbers when they go through `__torch_dispatch__`, fixing #74400 
* torch.ops now accept integers/floats anywhere Tensors were accepted, and they turn into wrapped scalar numbers
* PrimTorch refs now handle CPU scalar + CUDA correctly, fixes #76871

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

albanD commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -693,8 +693,24 @@ inline py::object toPyObject(IValue ivalue) {
+      auto scalar_type = tensor.scalar_type();
+      switch (scalar_type) {
+        case at::ScalarType::Bool: return py::cast(*tensor.data_ptr<bool>());
+        case at::ScalarType::Long: return py::cast(*tensor.data_ptr<int64_t>());
```
The int vs long namings are quite confusing... But looks good!

ngimel commented on torch/_refs/__init__.py:
```diff
@@ -441,6 +443,14 @@ def _convert(x):
     return tuple(map(lambda x: _convert(x), args))
 
 
+def _unwrap_cpu_scalars(a, b):
```
There's a million funny interactions here, and yeah if this unwrapping is done before type promotion it'll through it off (but it's done after, so that's fine). The reason for unwrapping user-provided tensors is that they stop beind 0d tensors after forced broadcasting #76871

albanD replied:
So we really want to keep the `elif` here? meaning that one will become a Number and the other will remain a 0D Tensor?

ngimel replied:
As it stands today (because prim implementation calls into `torch.add` and not into `torch.ops.aten.add` ) it doesn't matter, because python argparser will rewrap them again. But yeah, it's an interesting question, and if prim starts calling into aten, like it should, it would matter.

mruberry replied:
I think we should handle this by folding "scalar unwrapping" like this into a new type promotion wrapper I'm working on now, but no worries if this has an opportunity to land first

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -490,6 +502,11 @@ def _ref(a: Tensor, *, out: Optional[Tensor] = None) -> Tensor:
 
         return result
 
+    if aten_op is infer_aten_op:
```
Comment here please

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -467,7 +477,9 @@ def _maybe_resize_out(out: TensorLikeType, shape):
-def _make_elementwise_unary_reference(prim: Callable, *, type_promotion) -> Callable:
+infer_aten_op = object()
+
+def _make_elementwise_unary_reference(prim: Callable, *, type_promotion, aten_op=infer_aten_op) -> Callable:
```
We should document the constructor args

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -842,6 +881,7 @@ def float_power(
 # Sub is implemented specially because it has an alpha argument and
 #   is decomposed into multiple prims.
 # TODO: consider refactoring this with add impl
+@register_decomposition(torch.ops.aten.sub)
```
We should add a note for decompositions/aten registration and add the pattern to the runbook, too

mruberry reviewed (approved):
Pre-approving. I made a few inline notes about documentation. A mapping for registering ATen operators and decompositions might be more interesting than having it op-by-op, but nbd. Scalar stuff is fine to go in (as long as your base includes @ngimel's PR updating the tests with error inputs) -- my upcoming PR might tweak it but that's cool

ngimel commented on torch/_refs/__init__.py:
```diff
@@ -467,7 +477,9 @@ def _maybe_resize_out(out: TensorLikeType, shape):
 #
 
 # TODO: add type promotion support
```
nit: this comment becomes misplaced, and likely it's stale anyway.

ezyang committed:
Update on "Register PrimTorch refs as decompositions."


For the most part, PrimTorch refs have the same signature as their
ATen equivalents.  I modify most PrimTorch refs to register themselves
as decompositions, using the prim name they wrap to find the aten name
(except for a few cases where the prim/aten names mismatch).  There are
some exclusions, falling into one of two categories:

- The torch equivalent was already implemented as a CompositeImplicitAutograd
  decomposition in C++

- The ref doesn't support enough features (e.g., the real deal has more
  kwargs / overloads than are currently implemented)

PrimTorch refs are written as a single function that supports all
overloads, and this style is convenient for cases where we have a bundle
of overloads for what morally is a single overload with a Union type
on an argument (which we ought to have supported in
native_functions.yaml but blah); to support registering a single decomp
for all the overloads, we modify register_decomposition to register
to ALL overloads if you pass it an overload packet.  This is technically
BC breaking but no tests started failing because of it.

There's also two other major technical pieces to this PR:

* Wrapped number tensors get converted back into numbers when they go through `__torch_dispatch__`, fixing #74400 
* torch.ops now accept integers/floats anywhere Tensors were accepted, and they turn into wrapped scalar numbers
* PrimTorch refs now handle CPU scalar + CUDA correctly, fixes #76871

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Register PrimTorch refs as decompositions."


For the most part, PrimTorch refs have the same signature as their
ATen equivalents.  I modify most PrimTorch refs to register themselves
as decompositions, using the prim name they wrap to find the aten name
(except for a few cases where the prim/aten names mismatch).  There are
some exclusions, falling into one of two categories:

- The torch equivalent was already implemented as a CompositeImplicitAutograd
  decomposition in C++

- The ref doesn't support enough features (e.g., the real deal has more
  kwargs / overloads than are currently implemented)

PrimTorch refs are written as a single function that supports all
overloads, and this style is convenient for cases where we have a bundle
of overloads for what morally is a single overload with a Union type
on an argument (which we ought to have supported in
native_functions.yaml but blah); to support registering a single decomp
for all the overloads, we modify register_decomposition to register
to ALL overloads if you pass it an overload packet.  This is technically
BC breaking but no tests started failing because of it.

There's also two other major technical pieces to this PR:

* Wrapped number tensors get converted back into numbers when they go through `__torch_dispatch__`, fixing #74400 
* torch.ops now accept integers/floats anywhere Tensors were accepted, and they turn into wrapped scalar numbers
* PrimTorch refs now handle CPU scalar + CUDA correctly, fixes #76871

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Register PrimTorch refs as decompositions."


For the most part, PrimTorch refs have the same signature as their
ATen equivalents.  I modify most PrimTorch refs to register themselves
as decompositions, using the prim name they wrap to find the aten name
(except for a few cases where the prim/aten names mismatch).  There are
some exclusions, falling into one of two categories:

- The torch equivalent was already implemented as a CompositeImplicitAutograd
  decomposition in C++

- The ref doesn't support enough features (e.g., the real deal has more
  kwargs / overloads than are currently implemented)

PrimTorch refs are written as a single function that supports all
overloads, and this style is convenient for cases where we have a bundle
of overloads for what morally is a single overload with a Union type
on an argument (which we ought to have supported in
native_functions.yaml but blah); to support registering a single decomp
for all the overloads, we modify register_decomposition to register
to ALL overloads if you pass it an overload packet.  This is technically
BC breaking but no tests started failing because of it.

There's also two other major technical pieces to this PR:

* Wrapped number tensors get converted back into numbers when they go through `__torch_dispatch__`, fixing #74400 
* torch.ops now accept integers/floats anywhere Tensors were accepted, and they turn into wrapped scalar numbers
* PrimTorch refs now handle CPU scalar + CUDA correctly, fixes #76871

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Register PrimTorch refs as decompositions."


For the most part, PrimTorch refs have the same signature as their
ATen equivalents.  I modify most PrimTorch refs to register themselves
as decompositions, using the prim name they wrap to find the aten name
(except for a few cases where the prim/aten names mismatch).  There are
some exclusions, falling into one of two categories:

- The torch equivalent was already implemented as a CompositeImplicitAutograd
  decomposition in C++

- The ref doesn't support enough features (e.g., the real deal has more
  kwargs / overloads than are currently implemented)

PrimTorch refs are written as a single function that supports all
overloads, and this style is convenient for cases where we have a bundle
of overloads for what morally is a single overload with a Union type
on an argument (which we ought to have supported in
native_functions.yaml but blah); to support registering a single decomp
for all the overloads, we modify register_decomposition to register
to ALL overloads if you pass it an overload packet.  This is technically
BC breaking but no tests started failing because of it.

There's also two other major technical pieces to this PR:

* Wrapped number tensors get converted back into numbers when they go through `__torch_dispatch__`, fixing #74400 
* torch.ops now accept integers/floats anywhere Tensors were accepted, and they turn into wrapped scalar numbers
* PrimTorch refs now handle CPU scalar + CUDA correctly, fixes #76871

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Register PrimTorch refs as decompositions."


For the most part, PrimTorch refs have the same signature as their
ATen equivalents.  I modify most PrimTorch refs to register themselves
as decompositions, using the prim name they wrap to find the aten name
(except for a few cases where the prim/aten names mismatch).  There are
some exclusions, falling into one of two categories:

- The torch equivalent was already implemented as a CompositeImplicitAutograd
  decomposition in C++

- The ref doesn't support enough features (e.g., the real deal has more
  kwargs / overloads than are currently implemented)

PrimTorch refs are written as a single function that supports all
overloads, and this style is convenient for cases where we have a bundle
of overloads for what morally is a single overload with a Union type
on an argument (which we ought to have supported in
native_functions.yaml but blah); to support registering a single decomp
for all the overloads, we modify register_decomposition to register
to ALL overloads if you pass it an overload packet.  This is technically
BC breaking but no tests started failing because of it.

There's also two other major technical pieces to this PR:

* Wrapped number tensors get converted back into numbers when they go through `__torch_dispatch__`, fixing #74400 
* torch.ops now accept integers/floats anywhere Tensors were accepted, and they turn into wrapped scalar numbers
* PrimTorch refs now handle CPU scalar + CUDA correctly, fixes #76871

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Register PrimTorch refs as decompositions."


For the most part, PrimTorch refs have the same signature as their
ATen equivalents.  I modify most PrimTorch refs to register themselves
as decompositions, using the prim name they wrap to find the aten name
(except for a few cases where the prim/aten names mismatch).  There are
some exclusions, falling into one of two categories:

- The torch equivalent was already implemented as a CompositeImplicitAutograd
  decomposition in C++

- The ref doesn't support enough features (e.g., the real deal has more
  kwargs / overloads than are currently implemented)

PrimTorch refs are written as a single function that supports all
overloads, and this style is convenient for cases where we have a bundle
of overloads for what morally is a single overload with a Union type
on an argument (which we ought to have supported in
native_functions.yaml but blah); to support registering a single decomp
for all the overloads, we modify register_decomposition to register
to ALL overloads if you pass it an overload packet.  This is technically
BC breaking but no tests started failing because of it.

There's also two other major technical pieces to this PR:

* Wrapped number tensors get converted back into numbers when they go through `__torch_dispatch__`, fixing #74400 
* torch.ops now accept integers/floats anywhere Tensors were accepted, and they turn into wrapped scalar numbers
* PrimTorch refs now handle CPU scalar + CUDA correctly, fixes #76871

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge this
